### ISSUE TITLE
Change amp-bind to use custom events instead of MutationObservers

### ISSUE THREAD

**Issue by kmh287** [ampproject/amphtml#8673]

**Comment by kmh287** [comment]
@choumx Do you see a benefit to using events vs just having each element with dynamic content call some method directly on the bind service to inform it of updates?

**Comment by choumx** [comment]
Events are more flexible (can have multiple listeners) and simpler since you won't need to check for existence of the `amp-bind` extension.

**Comment by kmh287** [comment]
We're only creating these events for bind to listen to them, right? Also, won't `bindForDoc` just return a promise that never resolves if bind isn't installed?

**Comment by choumx** [comment]
Yes and yes. Abstractly, loose coupling is a better fit here.

**Comment by kmh287** [comment]
Which is an argument for continuing to use mutation observers, really. (Don't remember if we've discussed this) The current implementation relies on a MutationObserver to tell us which elements have been added or removed. Switching to custom events would require individual elements to fire events that include which elements have been added and removed, pushing bind business logic out to other elements and making developer error far more likely.
